fms/models/roberta.py

```diff
@@ -0,0 +1,335 @@
+        x = x + residual
+        x = self.ff_ln(x)
+
+        return x
+
+
+class RoBERTaHeadless(nn.Module):
+    def __init__(self, config: RoBERTaConfig):
+        super().__init__()
+        self.config = config
+
+        self.layers = nn.ModuleList(
+            [RoBERTaBlock(self.config) for _ in range(self.config.nlayers)]
+        )
+
+        self.embedding = nn.Embedding(self.config.src_vocab_size, self.config.emb_dim)
+        self.position_embedding = nn.Embedding(self.config.max_pos, self.config.emb_dim)
+
+        self.enc_norm = nn.LayerNorm(self.config.emb_dim, eps=self.config.norm_eps)
+
+        if self.config.p_dropout:
+            self.dropout = nn.Dropout(self.config.p_dropout)
+
+    def reset_params(self):
+        for layer in ["embedding", "position_embedding"]:
+            nn.init.normal_(
+                getattr(self, layer).weight,
+                mean=0.0,
+                std=self.config.emb_dim**-0.5,
+            )
+
+    def forward(
+        self,
+        x: torch.LongTensor,
+        mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        attn_algorithm: Optional[str] = None,
+    ):
+
+        if mask is None:
+            if x is None:
+                raise ValueError("cannot create a mask when x is None")
+            pad_id: int = self.config.pad_id
+            is_pad: torch.BoolTensor = x == pad_id
+            mask: torch.BoolTensor = is_pad.unsqueeze(-1) == is_pad.unsqueeze(-2)
+
+        x_emb = self.embedding(x)
+
+        # if pad_id exists
+        #   is_pad will be a BoolTensor
+        #   otherwise pad_id will not be taken into account
+        if self.config.pad_id is None:
+            is_pad = torch.zeros_like(x, dtype=bool, device=x.device)
+        else:
+            is_pad = x == self.config.pad_id
+
+        if position_ids is None:
+            position_ids = ((~is_pad).cumsum(1) - 1).clamp(min=0)
+
+        # look up position embeddings
+        position_out = self.position_embedding(position_ids)
+
+        # zero out the associated position embeddings
+        if self.config.pad_id is not None:
+            position_out = position_out.mul(~is_pad.unsqueeze(-1))
+
+        # perform absolute position embedding
+        x = x_emb + position_out
+
+        # layer norm
+        x = self.enc_norm(x)
+
+        # add dropout
+        if self.config.p_dropout:
+            x = self.dropout(x)
+
+        # layers
+        for layer in self.layers:
+            x = layer(x, mask=mask, attn_algorithm=attn_algorithm)
+
+        return x
+
+
+class RoBERTa(nn.Module):
+    def __init__(self, config: Optional[RoBERTaConfig] = None, **kwargs):
+
+        super(RoBERTa, self).__init__()
+        if config is not None:
+            self.config = config
+        else:
+            self.config = RoBERTaConfig()
+        self.config = self.config.updated(**kwargs)
+
+        self.base_model = RoBERTaHeadless(self.config)
```

Comment:
should we also accept a distribution_strategy, consistent with the llama model? that would make this compatible with options in the tuning scripts etc.

Reply:
Added to latest commit